### ISSUE TITLE
TouchableOpacity inside of TapGestureHandler

### ISSUE THREAD

**Issue by zsaraf** [software-mansion/react-native-gesture-handler#594]
I have many <TouchableOpacity /> components inside of one big container component that has a TapGestureHandler on it. For example:

```
import { TouchableOpacity } from 'react-native-gesture-handler'
public render() {
        return (
            <TapGestureHandler
                onHandlerStateChange={this._onSingleTap}
            >
                 <TouchableOpacity />
                 <TouchableOpacity />
            </TapGestureHandler>
        )
    }
```

The above example is a very basic one. In reality, my view is very complex and has many different `TouchableOpacity`s in addition to many `TextInput`s spanning across many wrapper components.
 
Currently, when tapping on one of the `TouchableOpacity`s, only `TapGestureRecognizer.onHandlerStateChange` is called. The exact opposite should happen, `TouchableOpacity.onPress` should get called and `TapGestureRecognizer.onHandlerStateChange` should NOT get called.

This is a very big issue for me, I would appreciate any help I can get! Has this been fixed in the upgrade branch?

Zach

**Comment by Chehow** [comment]
As an workaround you can write own TouchableOpacity based on TapGestureHandler and replace it in all places inside of your container. Anohter options - you can use react native PanResponder to handle this case (but if you have scroll you will face some issues on IOS).

**Comment by yolpsoftware** [comment]
@zsaraf did you find a solution for this?

**Comment by curiousdustin** [comment]
@Chehow , can you explain a bit more on how you can solve this with a custom TouchableOpacity? I have tried this concept, but when I nest them, tapping on a nested one, still triggers both the nested handler, and the container handler.

**Comment by yolpsoftware** [comment]
@curiousdustin basically we did something like this:

```
class TouchableOpacity extends Component<TouchableOpacityProps, any> {

    private opacityValue: Animated.Value<number> = new Animated.Value(1);
    private opacityClock = new Animated.Clock();

    gestureEvent(e) {
        if (e.nativeEvent.state == State.END) {
            this.opacityValue.setValue(.2);
            if (this.props.onPress) {
                this.props.onPress();
            }
        }
    }

    render() {
        return (
            <TapGestureHandler key={this.props.key} onHandlerStateChange={e => this.gestureEvent(e)}>
                <Animated.View style={[
                    this.props.style,
                    {
                        opacity: cond(
                            neq(this.opacityValue, 1),
                            block([
                                set(this.opacityValue, AnimationHelper.runTimingAnimation(this.opacityClock, this.opacityValue, 1, 300)),
                            ]),
                            1
                        )
                    }
                ]}>
                    {this.props.children}
                </Animated.View>
            </TapGestureHandler>
        )
    }
}
```

Not perfect, but seems to work fine so far.

**Comment by vldslvmgy** [comment]
is there any updates on this issue, facing the exact same one right now

**Comment by sa8ab** [comment]
Try putting each `TouchableOpacity` inside a `TabGestureHandler`

**Comment by shashwat3010** [comment]
> Try putting each `TouchableOpacity` inside a `TapGestureHandler`
> 
> ```
>             <TapGestureHandler
>                 onHandlerStateChange={this._onSingleTap}
>             >
>                 //.....
>                   <TapGestureHandler>
>                         <TouchableOpacity />
>                   <TapGestureHandler/>
>                   <TapGestureHandler>
>                         <TouchableOpacity />
>                   <TapGestureHandler/>
>                 //.....
>             </TapGestureHandler>
> ```

This worked for me

**Comment by andreev-danila** [comment]
> Try putting each `TouchableOpacity` inside a `TapGestureHandler`
> 
> ```
>             <TapGestureHandler
>                 onHandlerStateChange={this._onSingleTap}
>             >
>                 //.....
>                   <TapGestureHandler>
>                         <TouchableOpacity />
>                   <TapGestureHandler/>
>                   <TapGestureHandler>
>                         <TouchableOpacity />
>                   <TapGestureHandler/>
>                 //.....
>             </TapGestureHandler>
> ```

Thank you for this hack!

**Comment by yolpsoftware** [comment]
@sa8ab your workaround works well. Do you know why?

**Comment by javascripter** [comment]
I'm using `NativePressableOpacity` from this library as a workaround, which is based on TapGestureHandler.
https://github.com/mrousavy/react-native-pressable-opacity

**Comment by j-piasecki** [comment]
Hi! This should be fixed by setting `cancelsTouchesInView` prop on `TapGestureHandler` to `false`. This prop was added in https://github.com/software-mansion/react-native-gesture-handler/pull/1688 which will be included in the version 2.4.0 of Gesture Handler.

I will close this issue for now, and if the new prop doesn't solve your problem feel free to reopen it.

**Comment by stanwolverine** [comment]
@j-piasecki  By setting `cancelsTouchesInView: false`, now `onPress` callback of nested Touchables inside `TapGestureHandler` is called but `TapGestureHandler` is also registering that touch.

can we get a behaviour where `TapGestureHandler` does not register the touch but Touchables inside it do register the touch?
